File: leetcode/hard/2071_maxTaskAssign.ts
function maxTaskAssign(tasks: number[], workers: number[], pills: number, strength: number): number {
    let n = tasks.length;
    let m = workers.length;
    tasks.sort((a, b) => a - b);
    workers.sort((a, b) => a - b);


    const check = (mid: number): boolean => {
        let p: number = pills;
        let wQueue: number[] = new Array();
        let wId: number = m - 1;

        for (let i = mid - 1; i >= 0; --i) {
            while (wId >= workers.length - mid && workers[wId] + strength >= tasks[i]) {
                wQueue.unshift(workers[wId]);
                wId -= 1;
            }
            if (wQueue.length === 0) {
                return false;
            } else if (wQueue.at(-1) >= tasks[i]) {
                wQueue.pop();
            } else {
                if (!p) {
                    return false;
                }
                p -= 1;
                wQueue.shift();
            }
        }

        return true;
    };

    let left: number = 1
    let right: number = Math.min(m, n);
    let res: number = 0;
    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        if (check(mid)) {
            res = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return res;
};

/*

2 4 6 7 8 9 10; 1 3
1 3 6 9

*/

const test = () => {
    const params = [
        {
            input: { tasks: [10,10,10], workers: [7,7,10], pills: 1, strength: 3 },
            output: 2,
        },
        {
            input: { tasks: [3,2,1], workers: [0,3,3], pills: 1, strength: 1 },
            output: 3,
        },
        {
            input: { tasks: [5,4], workers: [0,0,0], pills: 1, strength: 5 },
            output: 1,
        },
        {
            input: { tasks: [10,15,30], workers: [0,10,10,10,10], pills: 3, strength: 10 },
            output: 2,
        },
    ];

    params.forEach(({input, output}) => {
        const { tasks, workers, pills, strength } = input;

        const result = maxTaskAssign(tasks, workers, pills, strength);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

File: leetcode/hard/2071_maxTaskAssign.py
from collections import deque
from typing import List
from json import dumps

class Solution:
    def maxTaskAssign(self, tasks: List[int], workers: List[int], pills: int, strength: int) -> int:
        n = len(tasks)
        m = len(workers)
        tasks.sort()
        workers.sort()

        def checkWorkers(taskId: int) -> bool:
            workersQueue = deque()
            workerId = m - 1
            pillCount = pills

            for i in range(taskId, -1, -1):
                while workerId >= 0 and workers[workerId] + strength >= tasks[i]:
                    workersQueue.appendleft(workers[workerId])
                    workerId -= 1

                if not workersQueue:
                    return False

                if workersQueue[-1] >= tasks[i]:
                    workersQueue.pop()
                elif pillCount == 0:
                    return False
                else:
                    workersQueue.popleft()
                    pillCount -= 1;


            return True

        res = 0
        left = 0
        right = min(n, m) - 1

        while (left <= right):
            middle = (left + right) // 2
            if (checkWorkers(middle)):
                res = middle + 1
                left = middle + 1
            else:
                right = middle - 1

        return res

def test ():
    params = [
        {
            'input': [[3,2,1], [0,3,3], 1, 1],
            'output': 3,
        },
        {
            'input': [[5,4], [0,0,0], 1, 5],
            'output': 1,
        },
        {
            'input': [[10,15,30], [0,10,10,10,10], 3, 10],
            'output': 2,
        },
    ]
    solution = Solution()

    for param in params:
        tasks, workers, pills, strength = param['input']
        result = solution.maxTaskAssign(tasks, workers, pills, strength)

        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()
